### PR TITLE
Cisco FTD | Create `security_event` fields from flattened `security` field

### PR DESCRIPTION
**Pull request by kcreddy**
## Proposed commit message
- WHAT: Parses out most fields from flattened `cisco.ftd.security` into new field `cisco.ftd.security_event`.
- WHY: Allows aggregation of fields under `cisco.ftd.security_event` which was not possible due to flattened `cisco.ftd.security` field.

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

Run this command for pipeline tests: 
`$ elastic-package stack down && elastic-package build && elastic-package stack up --version=8.10.3 -d -v && eval "$(elastic-package stack shellinit)" && elastic-package test pipeline --generate -v`

Expected files are now updated with new field `cisco.ftd.security_event` containing the fields listed from the linked issue.
```
Run pipeline tests for the package
2023/11/14 11:36:16 DEBUG GET https://127.0.0.1:5601/api/status
--- Test results for package: cisco_ftd - START ---
╭───────────┬─────────────┬───────────┬────────────────────────────────┬────────┬──────────────╮
│ PACKAGE   │ DATA STREAM │ TEST TYPE │ TEST NAME                      │ RESULT │ TIME ELAPSED │
├───────────┼─────────────┼───────────┼────────────────────────────────┼────────┼──────────────┤
│ cisco_ftd │ log         │ pipeline  │ test-asa-fix.log               │ PASS   │       5.95ms │
│ cisco_ftd │ log         │ pipeline  │ test-asa.log                   │ PASS   │ 141.360791ms │
│ cisco_ftd │ log         │ pipeline  │ test-dns.log                   │ PASS   │  28.012791ms │
│ cisco_ftd │ log         │ pipeline  │ test-filtered.log              │ PASS   │   3.302541ms │
│ cisco_ftd │ log         │ pipeline  │ test-firepower-management.log  │ PASS   │  31.007625ms │
│ cisco_ftd │ log         │ pipeline  │ test-ftd-endpoint-profile.log  │ PASS   │  49.369333ms │
│ cisco_ftd │ log         │ pipeline  │ test-ftd-fix.log               │ PASS   │      7.219ms │
│ cisco_ftd │ log         │ pipeline  │ test-ftd-inbound-outbound.log  │ PASS   │  20.108375ms │
│ cisco_ftd │ log         │ pipeline  │ test-intrusion.log             │ PASS   │  18.601208ms │
│ cisco_ftd │ log         │ pipeline  │ test-no-type-id.log            │ PASS   │  14.540875ms │
│ cisco_ftd │ log         │ pipeline  │ test-not-ip.log                │ PASS   │    4.35875ms │
│ cisco_ftd │ log         │ pipeline  │ test-sample.log                │ PASS   │  30.343333ms │
│ cisco_ftd │ log         │ pipeline  │ test-security-connection.log   │ PASS   │    10.5255ms │
│ cisco_ftd │ log         │ pipeline  │ test-security-file-malware.log │ PASS   │  32.704292ms │
│ cisco_ftd │ log         │ pipeline  │ test-security-malware-site.log │ PASS   │   7.476917ms │
╰───────────┴─────────────┴───────────┴────────────────────────────────┴────────┴──────────────╯
--- Test results for package: cisco_ftd - END   ---
Done
```

## Related issues

- Closes https://github.com/elastic/integrations/issues/5769

### CODE REVIEW

File: packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 1]
```diff
@@ -1348,7 +1348,10 @@ processors:
...
+        def dest_event = new HashMap();
+        def security_event_list = new ArrayList(['ac_policy', 'access_control_rule_action', 'access_control_rule_name', 'access_control_rule_reason', 'application_protocol', 'client', 'client_version', 'connection_duration', 'dns_query', 'dns_record_type', 'dns_response_type', 'dns_ttl', 'dst_ip', 'dst_port', 'egress_interface', 'egress_zone', 'file_action', 'file_count', 'file_direction', 'file_name', 'file_policy', 'file_sandbox_status', 'file_sha256', 'file_size', 'file_type', 'first_packet_second', 'http_referer', 'http_response', 'icmp_code', 'icmp_type', 'ingress_interface', 'ingress_zone', 'initiator_bytes', 'initiator_packets', 'nap_policy', 'prefilter_policy', 'protocol', 'referenced_host', 'responder_bytes', 'responder_packets', 'sha_disposition', 'spero_disposition', 'src_ip', 'src_port', 'ssl_actual_action', 'ssl_certificate', 'ssl_expected_action', 'ssl_flow_status', 'ssl_policy', 'ssl_rule_name', 'ssl_server_cert_status', 'ssl_server_name', 'ssl_session_id', 'ssl_ticket_id', 'ssl_version', 'sslurl_category', 'tunnel_or_prefilter_rule', 'uri', 'url', 'url_category', 'url_reputation', 'user', 'user_agent', 'web_application']);
```
Comment (efd6) [inline packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml:1352]: Can this be split over lines to reduce future diff noise?

File: packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -1370,6 +1378,126 @@ processors:
...
+      type: ip
+  - convert:
+      if: ctx._temp_?.cisco?.security_event?.src_port != null
+      tag: convert-security_event_src_port
+      field: _temp_.cisco.security_event.src_port
+      target_field: _temp_.cisco.security_event.src_port
+      type: integer
+
+  # Move out security fields out of flattened datatype. Related: https://github.com/elastic/integrations/issues/5769
+  # - set:
```
Comment (efd6) [inline packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml:1472]: Are there commented lines being retained intentionally?

File: packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -2430,11 +2558,11 @@ on_failure:
...
-  - remove:
-      field:
-        - _temp_
-        - _conf
-      ignore_missing: true
+  # - remove:
+  #     field:
+  #       - _temp_
+  #       - _conf
+  #     ignore_missing: true
```
Comment (efd6) [inline packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml:2565]: Are these being retained intentionally?

File: packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 1 above)
Comment (kcreddy) [reply packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml]: Updated as suggested

File: packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 2 above)
Comment (kcreddy) [reply packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml]: It wasn't intentional. It was a different approach I forgot to remove them. I removed them now.

File: packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 3 above)
Comment (kcreddy) [reply packages/cisco_ftd/data_stream/log/elasticsearch/ingest_pipeline/default.yml]: Removed the comments. It was supposed to be uncommented.

**Review by efd6 — COMMENTED** [review]
ISTM that this could be claimed to be a breaking change; if a user has an `@custom` pipeline that copies fields out of `cisco.ftd.security.*` and they are in the set of fields that are now moved to `cisco.ftd.security_event.*` then their custom pipeline will fail to result in documents whose structure downstream analysis may depend on.

**Comment by kcreddy** [comment]
>  if a user has an @custom pipeline that copies fields out of cisco.ftd.security.* and they are in the set of fields that are now moved to cisco.ftd.security_event.* then their custom pipeline will fail. 

I am not sure how we can handle this problem efficiently because `@custom` pipeline is always applied at the very end of the pipeline. Instead of moving, we can just copy the fields into `cisco.ftd.security_event.*`, thus keeping existing fields intact, but that creates a lot of duplicates. We could possibly inform the users that this upgrade will be a breaking change from the Fleet UI if thats how breaking changes are handled.

**Review by P1llus — APPROVED** [review]
> > if a user has an `@custom` pipeline that copies fields out of cisco.ftd.security.* and they are in the set of fields that are now moved to cisco.ftd.security_event.* then their custom pipeline will fail.
> 
> I am not sure how we can handle this problem efficiently because `@custom` pipeline is always applied at the very end of the pipeline. Instead of moving, we can just copy the fields into `cisco.ftd.security_event.*`, thus keeping existing fields intact, but that creates a lot of duplicates. We could possibly inform the users that this upgrade will be a breaking change from the Fleet UI if thats how breaking changes are handled.

I think we should ensure it is in the changelog title at least, as that is more prominent in the UI now compared to before, so that people know exactly what is going on.
We could keep the fields as well, but with that amount of fields it is a bit waste. I think we should just let @jamiehynds conclude on which approach we should take.

From my side its all good as long as its shown in the changelog and readme 👍

**Comment by kcreddy** [comment]
Updated the changelog title with `NOTE - This is a breaking change.`

File: packages/cisco_ftd/changelog.yml [hunk 4]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "2.21.0"
```
Comment (efd6) [inline packages/cisco_ftd/changelog.yml:2]: Since it's a breaking change, please bump major.

**Comment by narph** [comment]
cc @jamiehynds

**Comment by jamiehynds** [comment]
@kcreddy sorry I missed the original ping from Marius. As long as we're reflecting the breaking change in the changelog (and thus appearing in the UI when a user upgrades the integration), all good with me.

**Review by efd6 — APPROVED** [review]

**Pull request merged**